cunumeric/module.py

```diff
@@ -5327,6 +5336,630 @@ def nancumsum(
+@add_boilerplate("a")
+def nanprod(
+    a: ndarray,
+    axis: Any = None,
+    dtype: Any = None,
+    out: Union[ndarray, None] = None,
+    keepdims: bool = False,
+    initial: Optional[Union[int, float]] = None,
+    where: Any = True,
+) -> ndarray:
+    """
+    Return the product of array elements over a given axis treating
+    Not a Numbers (NaNs) as ones.
+
+    One is returned for slices that are all-NaN or empty.
+
+    Parameters
+    ----------
+    a : array_like
+        Input array.
+    axis : int, optional
+         Axis or axes along which the product is computed. The
+         default is to compute the product of the flattened array.
+    dtype : data-type, optional
+         The type of the returned array and of the accumulator in
+         which the elements are summed. By default, the dtype of a
+         is used. An exception is when a has an integer type with
+         less precision than the platform (u)intp. In that case,
+         the default will be either (u)int32 or (u)int64 depending
+         on whether the platform is 32 or 64 bits. For inexact
+         inputs, dtype must be inexact.
+    out : ndarray, optional
+        Alternate output array in which to place the result. The
+        default is None. If provided, it must have the same shape as
+        the expected output, but the type will be cast if necessary.
+        See Output type determination for more details. The casting of
+        NaN to integer can yield unexpected results.
+    keepdims : bool, optional
+        If this is set to True, the axes which are reduced are left in the
+        result as dimensions with size one. With this option, the result
+        will broadcast correctly against the input array.
+
+        If the default value is passed, then `keepdims` will not be
+        passed through to the `prod` method of sub-classes of
+        `ndarray`, however any non-default value will be.  If the
+        sub-class' method does not implement `keepdims` any
+        exceptions will be raised.
+    initial : scalar, optional
+        The starting value for this product. See `~cunumeric.ufunc.reduce` for
+        details.
+    where : array_like[bool], optional
+        Elements to include in the product. See `~cunumeric.ufunc.reduce` for
+        details.
+
+    Returns
+    -------
+    nanprod: ndarray, see `dtype` parameter above.
+        A new array holding the result is returned unless out is
+        specified, in which case it is returned.
+
+    See Also
+    --------
+    numpy.prod, numpy.isnan
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+
+    """
+
+    # Note: if the datatype of the input array is int and less
+    # than that of the platform int, then a convert task is launched
+    # in np.prod to take care of the type casting
+
+    # raise error for disallowed datatypes
+    disallowed_dtypes: list[np.dtype[Any]] = [
+        np.dtype(np.complex128),
+    ]
+    if a.dtype in disallowed_dtypes:
```

Comment:
```suggestion
    disallowed_dtypes: list[np.dtype[Any]] = [
        np.dtype(np.complex128),
    ]
    if a.dtype == np.complex128:
```
don't need to be fancy. even the error message is specific to that type.